File: 3016-minimum-number-of-pushes-to-type-word-ii/3016-minimum-number-of-pushes-to-type-word-ii.js
/**
 * @param {string} word
 * @return {number}
 */
var minimumPushes = function(word) {
    // 26-slot typed array allocated instantly on the stack
    const freq = new Int32Array(26);
    const len = word.length;
    
    // Direct character reading via charCodeAt to bypass array allocations
    for (let i = 0; i < len; i++) {
        freq[word.charCodeAt(i) - 97]++;
    }
    
    // Native fast sort
    freq.sort();
    
    let totalPushes = 0;
    let keyIndex = 0;
    
    // Scan backward from the highest frequencies
    for (let i = 25; i >= 0; i--) {
        if (freq[i] === 0) {
            break; // Stop immediately once all unique letters are processed
        }
        
        const pushCost = Math.floor(keyIndex / 8) + 1;
        totalPushes += freq[i] * pushCost;
        keyIndex++;
    }
    
    return totalPushes;
};

File: 3016-minimum-number-of-pushes-to-type-word-ii/3016-minimum-number-of-pushes-to-type-word-ii.java
import java.util.Arrays;

class Solution {
    public int minimumPushes(String word) {
        int[] freq = new int[26];
        int len = word.length();
        
        // FIX 1: Direct character scanning via charAt() to bypass heap allocation
        for (int i = 0; i < len; i++) {
            freq[word.charAt(i) - 'a']++;
        }

        // FIX 2: Fast primitive sort on the fixed 26-slot stack boundaries
        Arrays.sort(freq);

        int totalPushes = 0;
        int keyIndex = 0;

        // Step 3: Scan backwards (highest frequencies first) to assign push costs
        for (int i = 25; i >= 0; i--) {
            if (freq[i] == 0) {
                break; // Stop immediately once all existing characters are processed
            }
            
            // Push cost: first 8 get 1 push, next 8 get 2 pushes, etc.
            int pushCost = (keyIndex / 8) + 1;
            totalPushes += freq[i] * pushCost;
            keyIndex++;
        }

        return totalPushes;
    }
}
